libsolidity/codegen/CompilerUtils.cpp

```diff
@@ -1051,11 +1051,11 @@ void CompilerUtils::convertType(
 	case Type::Category::ArraySlice:
 	{
 		auto& typeOnStack = dynamic_cast<ArraySliceType const&>(_typeOnStack);
-		solUnimplementedAssert(
-			_targetType.dataStoredIn(DataLocation::CallData),
-			"Conversion from calldata slices to memory not yet implemented."
-		);
-		solAssert(_targetType == typeOnStack.arrayType(), "");
+		solAssert(_targetType.category() == Type::Category::Array, "");
+		auto const& targetArrayType = dynamic_cast<ArrayType const&>(_targetType);
+		solAssert(targetArrayType == *typeOnStack.arrayType().copyForLocation(targetArrayType.location(), false), "");
+		if (!_targetType.dataStoredIn(DataLocation::CallData))
+			return convertType(typeOnStack.arrayType(), _targetType);
 		solUnimplementedAssert(
```

Comment:
This assertion is generally relevant and will be true for all valid cases, won't it? I think it should move up before the early return, just to be sure.

Reply:
Yeah. Though maybe it should just be an ordinary assert because this seems to be covered by the type checker. `solUnimplementedAssert()` suggests that it could actually happen here.

On 0.7.5 slicing `uint[][] calldata` results in:
```
Error: Index range access is not supported for arrays with dynamically encoded base types.
```

Slicing `uint[][10] calldata` and `uint[10][10] calldata` results in:
```
Error: Index range access is only supported for dynamic calldata arrays.
```

Reply:
Yes, it indeed should be a regular assert - unfortunately at some point we realized that slicing for multi-dimensional arrays unfortunately will never be possible with the current way ABI encoding works.

Reply:
Multi-dimensional dynamic arrays or arrays in general?

Reply:
Yes - arrays of dynamic array address the inner array using a tail pointer that's relative to the "base reference", i.e. relative to the entire array. A slice doesn't have that reference anymore, unless we'd store it somewhere, which would make them three-stack-slot-types which would be rather silly... one can definitely work around it, but unclear, if the cost/benefit ratio of that both implementation-effort- and gas-wise will be justified in the end...

Reply:
dynamic arrays of static arrays could still work though - that part one might just classify as "unimplemented"...